clang/lib/Sema/SemaType.cpp

```diff
@@ -8511,7 +8514,7 @@ static void HandleHLSLParamModifierAttr(QualType &CurType,
     return;
   if (Attr.getSemanticSpelling() == HLSLParamModifierAttr::Keyword_inout ||
       Attr.getSemanticSpelling() == HLSLParamModifierAttr::Keyword_out)
-    CurType = S.getASTContext().getLValueReferenceType(CurType);
+    CurType = S.HLSL().getInoutParameterType(CurType);
```

Comment:
There are semantic restrictions here, right?  If nothing else, `CurType` can't be a reference type.

Reply:
HLSL doesn't allow references or pointers (at least not explicitly). We generate references in the AST in a few places, but any attempt to write or type-deduce one results in an error in Sema.

In theory, any legal type in HLSL can be converted to an inout parameter type.

Reply:
You can't reseat a reference, so I don't know what it would mean to pass one `inout`.  It should probably be invalid.

At any rate, if HLSL doesn't allow references, that's fine; please document that expectation with a comment, though (and maybe an assertion that we never see a reference here?).

Reply:
I put an assert in `SemaHLSL::getInoutParameterType` so that it will fire anywhere we try to adjust a type to the inout/out argument type.